crates/sui-core/src/authority_active/checkpoint_driver/mod.rs

```
@@ -0,0 +1,551 @@
+    // >2/3 of validators proposing a new checkpoint.
+    let next_proposal_sequence_number = highest_certificate_cert
+        .as_ref()
+        .map(|cert| cert.checkpoint.sequence_number + 1)
+        .unwrap_or(0);
+
+    // Collect proposals
+    let proposals: Vec<_> = final_state
+        .responses
+        .iter()
+        .filter_map(|(auth, proposal, _checkpoint)| {
+            if let Some(p) = proposal {
+                if p.0.checkpoint.sequence_number == next_proposal_sequence_number {
+                    return Some((*auth, p.clone()));
+                }
+            }
+            None
+        })
+        .collect();
+
+    Ok((highest_certificate_cert, proposals))
+}
+
+/// Download all checkpoints that are not known to us
+pub async fn sync_to_checkpoint<A>(
+    _active_authority: &ActiveAuthority<A>,
+    checkpoint_db: Arc<Mutex<CheckpointStore>>,
+    latest_known_checkpoint: CertifiedCheckpoint,
+) -> Result<(), SuiError>
+where
+    A: AuthorityAPI + Send + Sync + 'static + Clone,
+{
+    // Get out last checkpoint
+    let latest_checkpoint = checkpoint_db.lock().latest_stored_checkpoint()?;
+    let available_authorities: BTreeSet<_> = latest_known_checkpoint
+        .signatory_authorities()
+        .cloned()
+        .collect();
+
+    // Check if the latest checkpoint is merely a signed checkpoint, and if
+    // so download a full certificate for it.
+    if let Some(AuthenticatedCheckpoint::Signed(signed)) = &latest_checkpoint {
+        println!(
+            "Partial Sync ({:?}): {:?}",
+            _active_authority.state.name,
+            *signed.checkpoint.sequence_number()
+        );
+        let (past, _contents) = get_one_checkpoint(
+            _active_authority,
+            *signed.checkpoint.sequence_number(),
+            false,
+            &available_authorities,
+        )
+        .await?;
+
+        // NOTE: should we ignore the error here?
```

What is the intuition on why we could ignore the error here?

---

is the idea that if we fail here, we will simply go ahead and later process an higher checkpoint?

---

(in any case we should log an error)

---

Now logging the error. The issue is that the certificate may be for a later checkpoint than what we expect here.